Refactor roles

MyChakoTaco opened the pull request:
**Role Whitelisting**
There were big changes to move away from roles.py (whitelisting by code). Instead, the whitelisted roles will be under specific conditions:

1. bot has manage_roles permission
2. the white-listed role is assigned to the bot
3. the role has no permissions
4. the role is a subset of the bot's manage_roles role (this is done by role's position)

This allows the administrator (Ken) to add and remove roles without having to jump into the codebase and whitelist by roles.py

**Party Presets**
The party preset (size, emoji, imageURLs) are now moved away from roles.py and instead put into its own json file. Whenever parties are made, it'll check against this json file and apply preset data if there is a match.

**CreateRole and DeleteRole**
There are new role commands only Administrators can access.

They can use command .createrole to create a new role with no permissions, is automatically assigned to the bot, and put at the bottom of the role hierarchy. This satisfies the white-listed role conditions.

.deleterole deletes only game roles managed by the bot. I had considered having a fail-safe (type the command twice), but decided that Ken will know what he's doing.

MyChakoTaco committed:
Flake8 Fix

Use Flake8 Linter to prepare for future code changes

MyChakoTaco committed:
Modify Party Presets

- Add a JSON file to handle preset party data (redundant data will be removed along with roles.py)
- Have party class use JSON file
- Remove images folder (embed images only use HTTPS urls)

MyChakoTaco committed:
Integrate Role Handling

- Remove white-listing via roles.py
- Update valid game-role conditions by adding key conditions:
1) bot's roles include game-role
2) game role has no permissions (and will be treated as such in the future)
3) game role is in a subset of bot's manage_roles role

MyChakoTaco committed:
Create/Delete Role + Bugfix

- new roles_cog.py functionality:
1) .createrole - admin command to create a new role
2) .deleterole - admin command to delete an existing role

- bugfix party_cog.py where utility.isGamesRole was not updated to new definition

danny-tom commented on cogs/roles/roles_cog.py:
```diff
@@ -149,8 +155,8 @@ async def games(self, context):
-            if utility.isGamesRole(role):
-                games.append([str(role), str(len(role.members))])
+            if utility.isGamesRole(context.guild, self.bot, role):
+                games.append([str(role), str(len(role.members)-1)])
```
-1 to account for Charon?

MyChakoTaco replied:
Yup!

danny-tom commented:
Is removal of the game images intentional? We have to get it from outside source?

MyChakoTaco commented:
Oh I should've looked this up. We can use local files. Lol. I'll change it to use local files instead

Discordpy.readthedocs.io/en/latest/faq.html

danny-tom commented on utility/utility.py:
```diff
@@ -11,9 +12,41 @@ def getRole(roles, roleName):
+# returns if role is found in guild, has no permissions
+# and is manageable by the bot
+def isGamesRole(guild, bot, role):
+    botMember = getBotMember(guild, bot)
+
+    if botMember is None:
+        return False
```
Why do we need to find the bot user here? Isn't bot passed in so we have that to use? I am understanding that we want to use the bot user's roles in for the particular guild. Is there an easier way?

MyChakoTaco replied:
The utility class is purely functional. It does not have a class with a bot attribute like the cog classes do.

isGamesRole is used by both partycog and rolescog, so I had a tough time deciding to move isGamesRole out of the utility class.

Because of the nature of the utility class, it also does not know the context of its usage. That's why I had to pass in the guild and bot.

I would like to change the architecture, but I had a hard time figuring it out. Tbh I didn't give it much thought, and decided to fix it later when I'm up for it. If im going to make some fixes and you have an idea on architecture, I'm happy to make it work

danny-tom replied:
Ohhh Member is different than User in that it gives guild context

MyChakoTaco replied:
Yup! I learned that yesterday. Sorry I should've started with that! Yeah clientuser/user are related to profiles. Member is their context of the guild

danny-tom replied:
Could we use the Discord utility for simplicity
` botMember = discord.utils.find(lambda m: m == bot.user, guild.members)`

Essentially the same as what you're defining.

danny-tom commented on utility/utility.py:
```diff
@@ -11,9 +12,41 @@ def getRole(roles, roleName):
+    botRole = None
+    manageRolePermission = Permissions.none()
+    manageRolePermission.manage_roles = True
+    for r in botMember.roles:
```
Are roles ordered so that the roles with least permissions are first? So if someone mistakenly adds a role with admin into the list, we have botRole that includes that?

MyChakoTaco replied:
Roles are ordered from 0 to n, n being the highest priority role. When we add roles, they're always places at 0 and all other roles are bumped up. Permissions do not determine the order of roles, permissions are additive in nature, but roles with manage-roles can only affect roles with less priority.

I'm not sure if I get what you're saying, but I'll try to explain my theory. I'm finding the highest priority role of bot that allows it to manage role. I think it's up to the server managers to handle the bots permissions. Admin naturally comes with manage role and if it somehow gets put on the bot, Lord have mercy.

CasparHansen replied:
Are these checks necessary? According to [this part of the Discord developer documentation](https://discord.com/developers/docs/topics/permissions#permission-hierarchy), a bot can only grant roles that are lower permission than its highest role. If that's correct, the last two checks here should already be blocked by the API.

MyChakoTaco replied:
Good point. I'll eliminate some redundant checks. I was trying to make illustrative error feedback, but I can use except to catch permission errors instead.

MyChakoTaco committed:
Party Images re-added, Refactor roles and utility

Re-add party images
- Change party-class to use local URL
- Change party-cog to prepare party message properly instead of editting
- Add handling of local image file

roles_cog.py
- Rename isGamesRole to isGameRole for clarity
- Instead of checking for bot permissions, just handle try/except case.
- Fix some message formatting

utility.py
- Remove getBotMember function. (this can be handled with a one-liner from discord utils)
- refactor isGameRole logic to be more concise. We must keep the role positional and permission checks because not all commands will trigger manage_roles permissions (e.g. .games command)

MyChakoTaco commented:
@danny-tom @CasparHansen 

I made a new commit, let me know if I need to make any new changes.

Regarding the redundant checks... the discord API would correctly error out if our bot attempted to manage an inappropriate role. I added error handling to handle the resulting permission error.

Our utility function DOES need the redundant checks because we do need to know our game roles when we use commands that don't actually do any role management. An example is the .games or .party command. We don't want the bot to list roles it cannot actually manage in .games and we don't want .party to be able to mention roles it shouldn't be mentioning.

I was working on this for a couple hours trying to see if there was a good solution, but this is the best I could do so far. I would love some advice if you can see a good way to handle this issue.

danny-tom reviewed (approved):
Nice work!
